src/Operators/finitedifference.jl

```diff
@@ -1588,6 +1588,210 @@ end
+    Qⱼ⁻ = (ϕⱼᵗᵈ - ϕⱼᵐⁱⁿ) * dxⱼ
+    Rⱼ⁻ = (Pⱼ⁻ > stable_zero ? min(stable_one, Qⱼ⁻ / Pⱼ⁻) : stable_zero)
+
+    ϕⱼ₊₁ᵐᵃˣ = max(ϕⱼ, ϕⱼ₊₁, ϕⱼ₊₂, ϕⱼᵗᵈ, ϕⱼ₊₁ᵗᵈ, ϕⱼ₊₂ᵗᵈ)
+    ϕⱼ₊₁ᵐⁱⁿ = min(ϕⱼ, ϕⱼ₊₁, ϕⱼ₊₂, ϕⱼᵗᵈ, ϕⱼ₊₁ᵗᵈ, ϕⱼ₊₂ᵗᵈ)
+    Pⱼ₊₁⁺ = max(stable_zero, Aⱼ₊₁₂) - min(stable_zero, Aⱼ₊₃₂)
+    Qⱼ₊₁⁺ = (ϕⱼ₊₁ᵐᵃˣ - ϕⱼ₊₁ᵗᵈ) * dxⱼ₊₁
+    Rⱼ₊₁⁺ = (Pⱼ₊₁⁺ > stable_zero ? min(stable_one, Qⱼ₊₁⁺ / Pⱼ₊₁⁺) : stable_zero)
+    Pⱼ₊₁⁻ = max(stable_zero, Aⱼ₊₃₂) - min(stable_zero, Aⱼ₊₁₂)
+    Qⱼ₊₁⁻ = (ϕⱼ₊₁ᵗᵈ - ϕⱼ₊₁ᵐⁱⁿ) * dxⱼ₊₁
+    Rⱼ₊₁⁻ = (Pⱼ₊₁⁻ > stable_zero ? min(stable_one, Qⱼ₊₁⁻ / Pⱼ₊₁⁻) : stable_zero)
+
+    Cⱼ₊₁₂ = (Aⱼ₊₁₂ ≥ stable_zero ? min(Rⱼ₊₁⁺, Rⱼ⁻) : min(Rⱼ⁺, Rⱼ₊₁⁻))
+
+    return Cⱼ₊₁₂ * Aⱼ₊₁₂
+
+end
+
+stencil_interior_width(::FCTZalesak, A_space, Φ_space, Φᵗᵈ_space) =
+    ((-1, 1), (-half - 1, half + 1), (-half - 1, half + 1))
+
+@inline function stencil_interior(
+    ::FCTZalesak,
+    loc,
+    idx,
+    hidx,
+    A_space,
+    Φ_space,
+    Φᵗᵈ_space,
+)
+
+    center_space = axes(Φ_space)
+    face_space = axes(A_space)
+    # cell center variables
+    ϕⱼ₋₁ = getidx(Φ_space, loc, idx - half - 1, hidx)
+    ϕⱼ = getidx(Φ_space, loc, idx - half, hidx)
+    ϕⱼ₊₁ = getidx(Φ_space, loc, idx + half, hidx)
+    ϕⱼ₊₂ = getidx(Φ_space, loc, idx + half + 1, hidx)
+    # cell center variables
+    ϕⱼ₋₁ᵗᵈ = getidx(Φᵗᵈ_space, loc, idx - half - 1, hidx)
+    ϕⱼᵗᵈ = getidx(Φᵗᵈ_space, loc, idx - half, hidx)
+    ϕⱼ₊₁ᵗᵈ = getidx(Φᵗᵈ_space, loc, idx + half, hidx)
+    ϕⱼ₊₂ᵗᵈ = getidx(Φᵗᵈ_space, loc, idx + half + 1, hidx)
+    # cell face variables
+    Aⱼ₊₁₂ = Geometry.contravariant3(
+        getidx(A_space, loc, idx, hidx),
+        Geometry.LocalGeometry(face_space, idx, hidx),
+    )
+    Aⱼ₋₁₂ = Geometry.contravariant3(
+        getidx(A_space, loc, idx - 1, hidx),
+        Geometry.LocalGeometry(face_space, idx - 1, hidx),
+    )
+    Aⱼ₊₃₂ = Geometry.contravariant3(
+        getidx(A_space, loc, idx + 1, hidx),
+        Geometry.LocalGeometry(face_space, idx + 1, hidx),
+    )
+
+
+    dxⱼ = Geometry.LocalGeometry(center_space, idx - half, hidx).∂x∂ξ[1]
+    dxⱼ₊₁ = Geometry.LocalGeometry(center_space, idx + half, hidx).∂x∂ξ[1]
+
+
+    return Geometry.Contravariant3Vector(
+        fct_zalesak(
+            Aⱼ₋₁₂,
+            Aⱼ₊₁₂,
+            Aⱼ₊₃₂,
+            ϕⱼ₋₁,
+            ϕⱼ,
+            ϕⱼ₊₁,
+            ϕⱼ₊₂,
+            ϕⱼ₋₁ᵗᵈ,
+            ϕⱼᵗᵈ,
+            ϕⱼ₊₁ᵗᵈ,
+            ϕⱼ₊₂ᵗᵈ,
+            dxⱼ,
+            dxⱼ₊₁,
+        ),
+    )
+end
+
+boundary_width(
+    ::FCTZalesak,
+    ::FirstOrderOneSided,
+    A_space,
+    Φ_space,
+    Φᵗᵈ_space,
+) = 2
+
+@inline function stencil_left_boundary(
+    ::FCTZalesak,
+    bc::FirstOrderOneSided,
+    loc,
+    idx,
+    hidx,
+    A_space,
+    Φ_space,
+    Φᵗᵈ_space,
+)
+    face_space = axes(A_space)
+    @assert idx <= left_face_boundary_idx(face_space) + 1
+
+    return Geometry.Contravariant3Vector(zero(eltype(eltype(A_space))))
```

Comment:
Why two `eltype`s?

Reply:
@simonbyrne  helps me with this, I guess the first one returns a vector, and the second returns a scalar?